File: Bloc_de_gastos_personales.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
public class Bloc_de_gastos_personales {
    static ArrayList<Double> montos = new ArrayList<>();
    static ArrayList<String> categorias = new ArrayList<>();
    static ArrayList<String> descripciones = new ArrayList<>();
    static ArrayList<String> fechas = new ArrayList<>();
    static Scanner sc = new Scanner(System.in);
    //----------------------------------------------//
    //--|menu_principal_bloc_de_gastos_personales|--//
    //----------------------------------------------//
    static void mostrarMenu() {
        System.out.println();
        System.out.println("======================================");
        System.out.println("     BLOC DE GASTOS PERSONALES");
        System.out.println("======================================");
        System.out.println(" 1. Agregar gasto");
        System.out.println(" 2. Ver todos los gastos");
        System.out.println(" 3. Ver total gastado");
        System.out.println(" 4. Ver gastos por categoria");
        System.out.println(" 5. Eliminar un gasto");
        System.out.println(" 6. Ver resumen del mes");
        System.out.println(" 7. Salir");
        System.out.println("======================================");
    }
    //-------------------//
    //--|agregar_gasto|--//
    //-------------------//
    static void agregarGasto(double presupuesto) {
        if (montos.size() >= 100) {
            System.out.println("Limite de 100 gastos alcanzado.");
            return;
        }
        int numero = montos.size() + 1;
        System.out.println("-- Agregar Gasto #" + numero + " --");
        System.out.print("Monto ($): ");
        double monto = sc.nextDouble();
        sc.nextLine();
        System.out.print("Categoria (Comida/Transporte/Salud/Ocio/Otro): ");
        String categoria = sc.nextLine();
        System.out.print("Descripcion: ");
        String descripcion = sc.nextLine();
        System.out.print("Fecha (DD/MM/AAAA): ");
        String fecha = sc.nextLine();
        montos.add(monto);
        categorias.add(categoria);
        descripciones.add(descripcion);
        fechas.add(fecha);
        System.out.println("Gasto registrado correctamente.");
        double total = montos.stream().mapToDouble(Double::doubleValue).sum();
        if (total > presupuesto) {
            System.out.printf("ALERTA: Has superado tu presupuesto de $%.2f%n", presupuesto);
            System.out.printf("Total gastado hasta ahora: $%.2f%n", total);
        }
    }
    //--------------------------//
    //--|ver_todos_los_gastos|--//
    //--------------------------//
    static void verTodosLosGastos() {
        if (montos.isEmpty()) {
            System.out.println("No hay gastos registrados.");
            return;
        }
        System.out.println();
        System.out.println("N  | Fecha       | Categoria   | Monto      | Descripcion");
        System.out.println("-------------------------------------------------------------");
        for (int i = 0; i < montos.size(); i++) {
            System.out.printf("%-3d| %-12s| %-12s| $%-9.2f| %s%n",
                    i + 1, fechas.get(i), categorias.get(i),
                    montos.get(i), descripciones.get(i));
        }
    }
    //-----------------------//
    //--|ver_total_gastado|--//
    //-----------------------//
    static void verTotalGastado(double presupuesto) {
        if (montos.isEmpty()) {
            System.out.println("No hay gastos registrados.");
            return;
        }
        double total      = montos.stream().mapToDouble(Double::doubleValue).sum();
        double disponible = presupuesto - total;
        System.out.println();
        System.out.printf("Total gastado:     $%.2f%n", total);
        System.out.printf("Presupuesto:       $%.2f%n", presupuesto);
        if (disponible >= 0) {
            System.out.printf("Saldo disponible:  $%.2f%n", disponible);
        } else {
            System.out.printf("Exceso:            $%.2f%n", Math.abs(disponible));
        }
    }
    //------------------------------//
    //--|ver_gastos_por_categoria|--//
    //------------------------------//
    static void verGastosPorCategoria() {
        if (montos.isEmpty()) {
            System.out.println("No hay gastos registrados.");
            return;
        }
        System.out.print("Categoria a buscar: ");
        String catBuscar = sc.nextLine();
        int    encontrados = 0;
        double totalCat    = 0;
        System.out.println("\n--- Gastos en: " + catBuscar + " ---");
        for (int i = 0; i < montos.size(); i++) {
            if (categorias.get(i).equalsIgnoreCase(catBuscar)) {
                System.out.printf("%d | %s | $%.2f | %s%n",
                        i + 1, fechas.get(i), montos.get(i), descripciones.get(i));
                totalCat += montos.get(i);
                encontrados++;
            }
        }
        if (encontrados == 0) {
            System.out.println("No se encontraron gastos en esa categoria.");
        } else {
            System.out.println("Cantidad: " + encontrados);
            System.out.printf("Total:    $%.2f%n", totalCat);
        }
    }
    //-----------------------//
    //--|eliminar_un_gasto|--//
    //-----------------------//
    static void eliminarGasto() {
        if (montos.isEmpty()) {
            System.out.println("No hay gastos para eliminar.");
            return;
        }
        System.out.print("Numero del gasto a eliminar (1 a " + montos.size() + "): ");
        int num = sc.nextInt();
        sc.nextLine();
        if (num < 1 || num > montos.size()) {
            System.out.println("Numero invalido.");
            return;
        }
        int indice = num - 1;
        System.out.printf("Eliminado: $%.2f - %s%n", montos.get(indice), descripciones.get(indice));
        montos.remove(indice);
        categorias.remove(indice);
        descripciones.remove(indice);
        fechas.remove(indice);
        System.out.println("Gasto eliminado correctamente.");
    }
    //-------------------------//
    //--|ver_resumen_del_mes|--//
    //-------------------------//
    static void verResumenDelMes(double presupuesto) {
        if (montos.isEmpty()) {
            System.out.println("No hay gastos registrados.");
            return;
        }
        double total = montos.stream().mapToDouble(Double::doubleValue).sum();
        double mayor = montos.stream().mapToDouble(Double::doubleValue).max().getAsDouble();
        double menor = montos.stream().mapToDouble(Double::doubleValue).min().getAsDouble();
        double promedio = total / montos.size();
        System.out.println();
        System.out.println("======= RESUMEN DEL MES =======");
        System.out.println("Gastos registrados: " + montos.size());
        System.out.printf("Total gastado:      $%.2f%n", total);
        System.out.printf("Presupuesto:        $%.2f%n", presupuesto);
        System.out.printf("Gasto mayor:        $%.2f%n", mayor);
        System.out.printf("Gasto menor:        $%.2f%n", menor);
        System.out.printf("Promedio por gasto: $%.2f%n", promedio);
        System.out.println("\n--- Cantidad por categoria ---");
        Map<String, Integer> conteo = new HashMap<>();
        for (String cat : categorias) {
            conteo.put(cat, conteo.getOrDefault(cat, 0) + 1);
        }
        for (Map.Entry<String, Integer> entrada : conteo.entrySet()) {
            System.out.println(entrada.getKey() + ": " + entrada.getValue() + " gasto(s)");
        }
        System.out.println("================================");
    }
    public static void main(String[] args) {
        System.out.println("Bienvenido al Bloc de Gastos Personales");
        System.out.print("Define tu presupuesto mensual ($): ");
        double presupuesto = sc.nextDouble();
        sc.nextLine();
        while (true) {
            mostrarMenu();
            System.out.print("Elige una opcion: ");
            String opcion = sc.nextLine();
            switch (opcion) {
                case "1":
                    agregarGasto(presupuesto);
                    break;
                case "2":
                    verTodosLosGastos();
                    break;
                case "3":
                    verTotalGastado(presupuesto);
                    break;
                case "4":
                    verGastosPorCategoria();
                    break;
                case "5":
                    eliminarGasto();
                    break;
                case "6":
                    verResumenDelMes(presupuesto);
                    break;
                //------------------------------//
                //--|salir_del_menu_principal|--//
                //------------------------------//
                case "7":
                    System.out.println("Hasta luego. Cuida tus finanzas!");
                    sc.close();
                    return;
                default:
                    System.out.println("Opcion invalida. Intenta de nuevo.");
                    break;
            }
        }
    }
}

File: Bloc_de_gastos_personales.py
montos = []
categorias = []
descripciones = []
fechas = []
#----------------------------------------------#
#--|menu_principal_bloc_de_gastos_personales|--#
#----------------------------------------------#
def mostrar_menu():
    print("")
    print("======================================")
    print("     BLOC DE GASTOS PERSONALES")
    print("======================================")
    print(" 1. Agregar gasto")
    print(" 2. Ver todos los gastos")
    print(" 3. Ver total gastado")
    print(" 4. Ver gastos por categoria")
    print(" 5. Eliminar un gasto")
    print(" 6. Ver resumen del mes")
    print(" 7. Salir")
    print("======================================")
#-------------------#
#--|agregar_gasto|--#
#-------------------#
def agregar_gasto(presupuesto):
    if len(montos) >= 100:
        print("Limite de 100 gastos alcanzado.")
        return
    numero = len(montos) + 1
    print(f"-- Agregar Gasto #{numero} --")
    monto = float(input("Monto ($): "))
    categoria = input("Categoria (Comida/Transporte/Salud/Ocio/Otro): ")
    descripcion = input("Descripcion: ")
    fecha = input("Fecha (DD/MM/AAAA): ")
    montos.append(monto)
    categorias.append(categoria)
    descripciones.append(descripcion)
    fechas.append(fecha)
    print("Gasto registrado correctamente.")
    total = sum(montos)
    if total > presupuesto:
        print(f"ALERTA: Has superado tu presupuesto de ${presupuesto:.2f}")
        print(f"Total gastado hasta ahora: ${total:.2f}")
#-------------------------#
#--|ver_todo_los_gastos|--#
#-------------------------#
def ver_todos_los_gastos():
    if len(montos) == 0:
        print("No hay gastos registrados.")
        return
    print("")
    print("N  | Fecha       | Categoria   | Monto      | Descripcion")
    print("-------------------------------------------------------------")
    for i in range(len(montos)):
        print(f"{i+1}  | {fechas[i]}  | {categorias[i]:<12} | ${montos[i]:<9.2f} | {descripciones[i]}")
#-----------------------#
#--|ver_total_gastado|--#
#-----------------------#
def ver_total_gastado(presupuesto):
    if len(montos) == 0:
        print("No hay gastos registrados.")
        return
    total = sum(montos)
    disponible = presupuesto - total
    print("")
    print(f"Total gastado:     ${total:.2f}")
    print(f"Presupuesto:       ${presupuesto:.2f}")
    if disponible >= 0:
        print(f"Saldo disponible:  ${disponible:.2f}")
    else:
        print(f"Exceso:            ${abs(disponible):.2f}")
#------------------------------#
#--|ver_gastos_por_categoria|--#
#------------------------------#
def ver_gastos_por_categoria():
    if len(montos) == 0:
        print("No hay gastos registrados.")
        return
    cat_buscar = input("Categoria a buscar: ")
    encontrados = 0
    total_cat = 0
    print(f"\n--- Gastos en: {cat_buscar} ---")
    for i in range(len(montos)):
        if categorias[i].lower() == cat_buscar.lower():
            print(f"{i+1} | {fechas[i]} | ${montos[i]:.2f} | {descripciones[i]}")
            total_cat += montos[i]
            encontrados += 1
    if encontrados == 0:
        print("No se encontraron gastos en esa categoria.")
    else:
        print(f"Cantidad: {encontrados}")
        print(f"Total:    ${total_cat:.2f}")
#-----------------------#
#--|eliminar_un_gasto|--#
#-----------------------#
def eliminar_gasto():
    if len(montos) == 0:
        print("No hay gastos para eliminar.")
        return
    num = int(input(f"Numero del gasto a eliminar (1 a {len(montos)}): "))
    if num < 1 or num > len(montos):
        print("Numero invalido.")
        return
    indice = num - 1
    print(f"Eliminado: ${montos[indice]:.2f} - {descripciones[indice]}")
    montos.pop(indice)
    categorias.pop(indice)
    descripciones.pop(indice)
    fechas.pop(indice)
    print("Gasto eliminado correctamente.")
#-------------------------#
#--|ver_resumen_del_mes|--#
#-------------------------#
def ver_resumen_del_mes(presupuesto):
    if len(montos) == 0:
        print("No hay gastos registrados.")
        return
    total    = sum(montos)
    mayor    = max(montos)
    menor    = min(montos)
    promedio = total / len(montos)
    print("")
    print("======= RESUMEN DEL MES =======")
    print(f"Gastos registrados: {len(montos)}")
    print(f"Total gastado:      ${total:.2f}")
    print(f"Presupuesto:        ${presupuesto:.2f}")
    print(f"Gasto mayor:        ${mayor:.2f}")
    print(f"Gasto menor:        ${menor:.2f}")
    print(f"Promedio por gasto: ${promedio:.2f}")
    print("\n--- Cantidad por categoria ---")
    conteo = {}
    for cat in categorias:
        conteo[cat] = conteo.get(cat, 0) + 1
    for cat, cantidad in conteo.items():
        print(f"{cat}: {cantidad} gasto(s)")
    print("================================")
print("Bienvenido al Bloc de Gastos Personales")
presupuesto = float(input("Define tu presupuesto mensual ($): "))
while True:
    mostrar_menu()
    opcion = input("Elige una opcion: ")
    if opcion == "1":
        agregar_gasto(presupuesto)
    elif opcion == "2":
        ver_todos_los_gastos()
    elif opcion == "3":
        ver_total_gastado(presupuesto)
    elif opcion == "4":
        ver_gastos_por_categoria()
    elif opcion == "5":
        eliminar_gasto()
    elif opcion == "6":
        ver_resumen_del_mes(presupuesto)
    #------------------------------#
    #--|salir_del_menu_principal|--#
    #------------------------------#
    elif opcion == "7":
        print("Hasta luego. Cuida tus finanzas!")
        break
    else:
        print("Opcion invalida. Intenta de nuevo.")